tests/backup_helper.go

```diff
@@ -2932,6 +3028,319 @@ func ValidateBackup(ctx context1.Context, backupName string, orgID string, sched
 	return err
 }
 
+// ValidateBackupWithPartialSuccess validates a backup's spec's objects (resources) and volumes. resourceTypesFilter can be used to select specific types to validate (nil means all types). This function must be called after switching to the context on which `scheduledAppContexts` exists. Cluster level resources aren't validated.
+func ValidateBackupWithPartialSuccess(ctx context1.Context, backupName string, orgID string, scheduledAppContexts []*scheduler.Context, resourceTypesFilter []string, failedVolumes []string) error {
+	var backupInspectResponse *api.BackupInspectResponse
+	log.InfoD("Validating backup [%s] in org [%s]", backupName, orgID)
+	log.Infof("Obtaining backup info for backup [%s]", backupName)
+	backupDriver := Inst().Backup
+	backupUid, err := backupDriver.GetBackupUID(ctx, backupName, orgID)
+	if err != nil {
+		return fmt.Errorf("GetBackupUID Err: %v", err)
+	}
+	backupInspectRequest := &api.BackupInspectRequest{
+		Name:  backupName,
+		Uid:   backupUid,
+		OrgId: orgID,
+	}
+	backupStatusCheck := func() (interface{}, bool, error) {
+		backupInspectResponse, err = backupDriver.InspectBackup(ctx, backupInspectRequest)
+		if err != nil {
+			return "", false, fmt.Errorf("InspectBackup Err: %v", err)
+		}
+		backupStatus := backupInspectResponse.GetBackup().GetStatus().Status
+		if backupStatus == api.BackupInfo_StatusInfo_PartialSuccess {
+			return "", false, nil
+		}
+		return "", true, fmt.Errorf("validate backup requires backup [%s] to have a status of PartialSuccess,got -%v", backupName, backupStatus)
+	}
+	_, err = DoRetryWithTimeoutWithGinkgoRecover(backupStatusCheck, MaxWaitPeriodForBackupCompletionInMinutes*time.Minute, 30*time.Second)
+	if err != nil {
+		return err
+	}
+	// Check size of backup taken is non-zero
+	resp, err := Inst().Backup.InspectBackup(ctx, backupInspectRequest)
+	if err != nil {
+		return err
+	}
+	Volumes := resp.GetBackup().GetVolumes()
+	if len(Volumes) > 0 {
+		for _, volume := range Volumes {
+			skipValidation := false
+			isFailedVolume := false
+			for _, failed := range failedVolumes {
+				if volume.Pvc == failed {
+					isFailedVolume = true
+					if volume.Status.Status == api.BackupInfo_StatusInfo_Failed {
```

Comment:
This check wont be accurate always, if the backup is partial.
instead of going for checking vol status with backup status.

The check can be like.
1. Get the backed up volumes. Make sure whatever we sent as failed volumes are actually failed.
2. And whatever not within failed volumes should be success.

The can be separate check outside this loop.

Reply:
@ak-px please advise, i'm just comparing the volume status with either 'Success' or 'Failed' and not comparing with the backup status. The check was successful when i tested.

Reply:
The best way is to create a map of expected failed volumes where the key = volume name and value = true.
Then you iterate over the volumes given by the backup object and check it against this map. Something like this

```
failedVolumeMap := make(map[string]bool)
	for _, failedVol := range failedVolumes {
		failedVolumeSet[failedVol] = true
	}

for _, volume := range volumes {
		isFailedVolume := failedVolumeSet[volume.Pvc]

		if isFailedVolume {
			if volume.Status.Status != api.BackupInfo_StatusInfo_Failed {
				log.Warnf("Volume PVC: %s is in failedVolumes but not actually failed", volume.Pvc)
			} else {
				log.Infof("Skipping validation for failed volume PVC: %s", volume.Pvc)
				continue
			}
		} else if volume.Status.Status != api.BackupInfo_StatusInfo_Success {
			log.Errorf("Volume PVC: %s is not in failedVolumes and does not have success status", volume.Pvc)
			return fmt.Errorf("volume PVC: %s is not in failedVolumes and does not have success status", volume.Pvc)
		}
```

Reply:
Always avoid nested loops from here on